sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/HttpTransportClient.java

```diff
@@ -863,15 +863,19 @@ static String getHeader(String[] names, String[] values, String name) {
                                 break;
                             } else {
                                 // Have the request URL in the exception message for debugging purposes.
-                                exception = new GoneException(
+                                GoneException goneExceptionFromService = new GoneException(
                                         String.format(
                                                 RMResources.ExceptionMessage,
                                                 RMResources.Gone),
                                         response.headers(),
                                         request.uri());
+                                goneExceptionFromService.setIsBasedOn410ResponseFromService();
```

Comment:
Should we be more specific here? only when substatus code  = 0, then retry the write operation.

Reply:
Synced offline.

Reply:
To elaborate what Annie and I discussed offline: Setting the flag indicating that a 410 results in a GoneException because of retrieving a 410 from the service is intentional. We special case certain sub status codes because we need to change the business logic upstream to refresh caches or route differently etc. - there might be additional sub status codes in the future which require special handling and we can add new exception types then.
This new flag in the GoneException simply indicates whether a 410 was retrieved form the service - it is used right now to determine whether from an idempotency perspective it is safe to rtery (which is true for any 410 retrieved form the service - we simply don't need the flag in the other special exceptions used for 410 with different substatuscode like PartitionSplitException etc.) - So not constraining setting teh flag to just sub status code 0 is intentional - and if the backend should introduce any additional sub status code it would still continue to be safe to retry from an idempotency perspective - so there is no concern there that would force us to special case substatus code 0 etc.
I disagree that using a different excetion type would make sense for this case - the GoneException indicating that we retrieved a 410 Gone status code was originally made for this case where we actually get a 410 response. If we would want to signal the different cases for which we use GoneException right now - all the transient GoneExceptions that we generate client-side (because of transport exceptions, because we can't get address for primary replica etc.) should have their own exception types. IMO the benefit is very small (it would help to avoid some confusion - but not much more) and the refactoring would be significant. I am not in favor of doing it, Annie agreed in offline discussion and if anyone thinks this should happen I think it should happen in a separate PR to do the refactoring of all the transient client-side generated GoneExceptions.